cvpaperchallenge/Crux
# Enhancement/add rag basics

## Issue URL

NA

## Change overview

- Update old Python libraries to the latest (including ones for dev)
- Implement the basic system for RAG-based chat system
- Add a Dockerfile to deploy the backend to AWS

## How to test

### 1. Prerequisite

1. Prepare the API keys for [OpenAI API](https://platform.openai.com/overview), [Mathpix OCR](https://mathpix.com/ocr), and [Qdrant Cloud](https://qdrant.tech/documentation/cloud/quickstart-cloud/)
2. Create your own `environments/.env` referring `environments/.env.sample` and specify the keys above as environment variables
3. Download the OCR-ed paper data from [here](https://github.com/cvpaperchallenge/Crux/files/14826455/test_papers.zip) and place as below.

    ```bash
    ├── data/
    │    │
    │    ├── test_papers/
    │    │    │
    │    │    ├── 01_Action-Conditioned_3D_Human_Motion_Synthesis_with_Transformer_VAE/
    │    │    │    │
    │    │    │    ├── 01_Action-Conditioned_3D_Human_Motion_Synthesis_with_Transformer_VAE_mathpix.txt
    │    │    │    │
    │    │    │    └── 01_Action-Conditioned_3D_Human_Motion_Synthesis_with_Transformer_VAE.pdf
    │    │    │
    │    │    └── 03_Generative_Adversarial_Graph_Convolutional_Networks_for_Human_Action_Synthesis/
    │    │         │
    │    │         ├── 03_Generative_Adversarial_Graph_Convolutional_Networks_for_Human_Action_Synthesis_mathpix.txt
    │    │         │
    │    │         └── 03_Generative_Adversarial_Graph_Convolutional_Networks_for_Human_Action_Synthesis.pdf
    │    │
    │    └ README.md
    ```

### 2. Local Test

#### Due to modifications to the Dockerfile, it is necessary to restart the container once.

1. Remove the existing containers

```bash
# Move to the directory that has `docker-compose.yaml`
~/Crux$ cd environments/cpu

# Remove the existing docker containers
~/Crux/environments/cpu$ docker compose down
```

4. Boot up containers without using cache

```bash
# Re-build docker images without using cache
~/Crux/environments/cpu$ docker compose build --no-cache

# Boot up docker containers
~/Crux/environments/cpu$ docker compose up -d
```

#### All the steps below should be processed inside the container

5. Update Python libraries inside the container

```bash
# Update the poetry.lock and install the updated python libraries
~/crux-backend$ poetry update
```

6. Run the backend server

```bash
~/crux-backend$ poetry run make run-backend
poetry run gunicorn 'src.main:app' -k uvicorn.workers.UvicornWorker -b 0.0.0.0:8000 -t 6000 --log-level info
[2024-04-01 02:51:06 +0000] [39523] [INFO] Starting gunicorn 21.2.0
[2024-04-01 02:51:06 +0000] [39523] [INFO] Listening at: http://0.0.0.0:8000 (39523)
[2024-04-01 02:51:06 +0000] [39523] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2024-04-01 02:51:06 +0000] [39539] [INFO] Booting worker with pid: 39539
[2024-04-01 02:51:06 +0000] [39539] [INFO] Started server process [39539]
[2024-04-01 02:51:06 +0000] [39539] [INFO] Waiting for application startup.
[2024-04-01 02:51:06 +0000] [39539] [INFO] Application startup complete.
```

7. Check if the RAG-based chat logic works properly on http://localhost:8000/docs

    path parameters (`user_id` and `conversation_id`) are to be used to treat each user's conversation as different and to provide different chat history
    ![image](https://github.com/cvpaperchallenge/Crux/assets/61688218/33ad5be0-566b-4109-a6a9-9f6ecb058098)

    If you get the responses below, the logic works correctly.
    ![image](https://github.com/cvpaperchallenge/Crux/assets/61688218/e2232e8d-f09a-4c0b-9dd1-f3b02ba7508d)

### 3. Test on AWS

I deployed it as a serverless application to AWS Lambda. You can verify its operation by sending a request to the Lambda function's URL. Calling the lambda from a function URL requires the signature with AWS Signature Version 4, but the signature generation process is complicated and takes much time.  Thus, using tools like Postman is recommended [link](https://docs.aws.amazon.com/ja_jp/lambda/latest/dg/urls-invocation.html#urls-invocation-basics)

The command below is one example of calling a lambda function with the basic curl command. [link](https://gist.github.com/mryhryki/28fcd54e8a8cdffb78462d171ce48b27)
*I **DID NOT** check the below command works without problems as I used Postman.

```bash
curl --location '<LAMBDA_FUNCTION_URL>/chat/1/1' \
--header 'Content-Type: application/json' \
--aws-sigv4 "aws:amz:<REGION>:lambda" \
--user "<AWS_ACCESS_KEY_ID>:<AWS_SECRET_ACCESS_KEY>" \
--data '{
  "query": "motion synthesisとはどのようなタスクを指すのでしょうか?",
  "qdrant_type": "cloud",
  "openai_api_key": "<OPENAI_API_KEY>",
  "qdrant_cloud_url": "<QDRANT_CLOUD_URL>",
  "qdrant_api_key": "<QDRANT_API_KEY>"
}'
```

---

The following process summarizes the steps to deploy a serverless application using AWS Lambda functions. Although it's possible to test using the above command, I've briefly summarized it here for sharing and recording purposes.

1. Install AWS CLI on the host PC. [link](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html)
*all commands on the following step should be executed on your host PC, not inside the container

2. Authenticate the Docker client against the Amazon ECR registry (issue a 12-hour valid token and perform docker login, it's necessary to execute this again after 12 hours).

```bash
~$ aws ecr get-login-password --region <REGION> | docker login --username AWS --password-stdin <AWS_ACCOUNT_ID>.dkr.ecr.<REGION>.amazonaws.com/<ECR_REPOSITORY_NAME>
```

3. Build `environments/deploy/Dockerfile.backend` to create a Docker image.

```bash
~/Crux$ docker build -f environments/deploy/Dockerfile.backend -t <AWS_ACCOUNT_ID>.dkr.ecr.<REGION>.amazonaws.com/<ECR_REPOSITORY_NAME>:<ARBITRARY_TAG> .
```

4. Upload the created image to Amazon ECR.

```bash
~/Crux$ docker push <AWS_ACCOUNT_ID>.dkr.ecr.<REGION>.amazonaws.com/<ECR_REPOSITORY_NAME>:<ARBITRARY_TAG>
```

5. Create a function from the container image uploaded to Amazon ECR on the AWS Lambda page. [link](https://qiita.com/kyamamoto9120/items/f1cda89ffc7cb5254f17#232-%E9%96%A2%E6%95%B0%E3%82%92%E4%BD%9C%E6%88%90)

6. Change the timeout settings of the Lambda function and the memory of the resources to be provisioned as shown in the photo.

    ![スクリーンショット 2024-04-03 164241](https://github.com/cvpaperchallenge/Crux/assets/61688218/1509dd13-9587-4a4c-9996-7287ab3adccb)

7. Generate the function URL for the created function. [link](https://docs.aws.amazon.com/lambda/latest/dg/urls-configuration.html#create-url-console)
*This time, since I wanted only authorized persons to access it, I set the Auth Type to AWS_IAM.

## Note for reviewers

> [!Note]
> The CD process using GitHub Actions has not been implemented yet, but there are plans to introduce it in the future.

> [!Important]
> See the "Note for reviewers" section of #30
